### PR TITLE
Bhavyssh/add UT

### PR DESCRIPTION
**Pull request by bhavya2109sharma**
*Issue #, if available:*

- We have unit tests for Marquee packages in SMD but we still have some external packages which either have only import test or no test.

*Description of changes:*

- Adding UT tests for the packages which don’t have test at all or have only export test cases.
- Test cases added and modified for the following packages -
    1. Jupyter-collaboration
    2. jupyterlab-git
    3. jupyterlab-lsp
    4. notebook
    5. python-lsp-server
    6. langchain
    7. fastapi
    8. uvicorn
    9. torchvision
    10. scikit-learn
    11. jinja2
    12. ipywidgets
    13. supervisor
    14. jupyter-scheduler
    15. jupyter-server-proxy
    16. tf-keras
    17. py-xgboost-cpu
    18. py-xgboost-gpu
    
 **Test Results for Image v2.3.3-CPU**
```
[gw34] [  2%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_cpu[jinja2.test.Dockerfile-required_packages34] 
[gw15] [ 22%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_cpu[jupyterlab-lsp.test.Dockerfile-required_packages15] 
[gw16] [ 34%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_cpu[python-lsp-server.test.Dockerfile-required_packages16] 
[gw26] [ 40%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_cpu[langchain.test.Dockerfile-required_packages26] 
[gw37] [ 47%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_cpu[scikit-learn.test.Dockerfile-required_packages37] 
[gw39] [ 50%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_cpu[ipywidgets.test.Dockerfile-required_packages40] 
[gw36] [ 52%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_cpu[fastapi.test.Dockerfile-required_packages36] 
[gw40] [ 56%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_cpu[jupyter-server-proxy.test.Dockerfile-required_packages39] 
[gw35] [ 59%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_cpu[uvicorn.test.Dockerfile-required_packages35] 
[gw41] [ 65%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_cpu[supervisor.test.Dockerfile-required_packages41] 
[gw38] [ 68%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_cpu[jupyter-scheduler.test.Dockerfile-required_packages38] 
[gw18] [ 72%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_cpu[notebook.test.Dockerfile-required_packages18] 
[gw23] [ 75%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_cpu[jupyterlab-git.test.Dockerfile-required_packages23] 
[gw13] [ 77%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_cpu[jupyter-collaboration.test.Dockerfile-required_packages13] 
[gw43] [ 88%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_cpu[xgboost-cpu.test.Dockerfile-required_packages43] 
```

 **Test Results for Image v2.4.1-CPU**
```
[gw15] [ 23%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_cpu[jupyterlab-lsp.test.Dockerfile-required_packages15] 
[gw40] [ 29%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_cpu[ipywidgets.test.Dockerfile-required_packages40] 
[gw37] [ 31%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_cpu[fastapi.test.Dockerfile-required_packages37] 
[gw26] [ 38%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_cpu[langchain.test.Dockerfile-required_packages26] 
[gw36] [ 40%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_cpu[jinja2.test.Dockerfile-required_packages35] 
[gw47] [ 48%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_cpu[scikit-learn.test.Dockerfile-required_packages46] 
[gw41] [ 53%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_cpu[supervisor.test.Dockerfile-required_packages41] 
[gw40] [ 56%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_cpu[jupyter-server-proxy.test.Dockerfile-required_packages39] 
[gw16] [ 57%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_cpu[python-lsp-server.test.Dockerfile-required_packages16] 
[gw35] [ 59%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_cpu[uvicorn.test.Dockerfile-required_packages36] 
[gw39] [ 63%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_cpu[jupyter-scheduler.test.Dockerfile-required_packages39] 
[gw18] [ 65%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_cpu[notebook.test.Dockerfile-required_packages18] 
[gw44] [ 68%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_cpu[jupyter-collaboration.test.Dockerfile-required_packages44] 
[gw22] [ 72%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_cpu[jupyterlab-git.test.Dockerfile-required_packages23] 
gw0] [80%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_cpu[xgboost-cpu.test.Dockerfile-required_packages43] 
[gw38] [ 91%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_cpu[torchvision.test.Dockerfile-required_packages38

```
 **Test Results for Image v3.0.0-CPU**
```
[gw15] [ 29%] PASSED test/test__dockerfile_based_harness.py::test_dockerfiles_for_cpu[jupyterlab-lsp.test.Dockerfile-required_packages15] 
[gw40] [ 34%] PASSED test/test__dockerfile_based_harness.py::test_dockerfiles_for_cpu[ipywidgets.test.Dockerfile-required_packages40] 
[gw37] [ 36%] PASSED test/test__dockerfile_based_harness.py::test_dockerfiles_for_cpu[fastapi.test.Dockerfile-required_packages36] 
[gw34] [ 40%] PASSED test/test__dockerfile_based_harness.py::test_dockerfiles_for_cpu[jinja2.test.Dockerfile-required_packages34] 
[gw38] [ 50%] PASSED test/test__dockerfile_based_harness.py::test_dockerfiles_for_cpu[scikit-learn.test.Dockerfile-required_packages37] 
[gw41] [ 54%] PASSED test/test__dockerfile_based_harness.py::test_dockerfiles_for_cpu[supervisor.test.Dockerfile-required_packages41] 
[gw17] [ 59%] PASSED test/test__dockerfile_based_harness.py::test_dockerfiles_for_cpu[python-lsp-server.test.Dockerfile-required_packages16] 
[gw36] [ 61%] PASSED test/test__dockerfile_based_harness.py::test_dockerfiles_for_cpu[uvicorn.test.Dockerfile-required_packages35] 
[gw39] [ 65%] PASSED test/test__dockerfile_based_harness.py::test_dockerfiles_for_cpu[jupyter-server-proxy.test.Dockerfile-required_packages39] 
[gw35] [ 68%] PASSED test/test__dockerfile_based_harness.py::test_dockerfiles_for_cpu[jupyter-scheduler.test.Dockerfile-required_packages38] 
[gw18] [ 70%] PASSED test/test__dockerfile_based_harness.py::test_dockerfiles_for_cpu[notebook.test.Dockerfile-required_packages18] 
[gw23] [ 72%] PASSED test/test__dockerfile_based_harness.py::test_dockerfiles_for_cpu[jupyterlab-git.test.Dockerfile-required_packages23]
[gw0] [77%] PASSED test/test__dockerfile_based_harness.py::test_dockerfiles_for_cpu[jupyter-collaboration.test.Dockerfile-required_packages13] 
[gw43] [ 90%] PASSED test/test__dockerfile_based_harness.py::test_dockerfiles_for_cpu[xgboost-cpu.test.Dockerfile-required_packages43]
[gw38] [ 93%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_cpu[torchvision.test.Dockerfile-required_packages38
```

**Test results for v2.3.3-GPU**
```
[gw7] [ 10%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_gpu[jupyterlab-lsp.test.Dockerfile-required_packages15] 
[gw7] [ 13%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_gpu[notebook.test.Dockerfile-required_packages18] 
[gw1] [ 21%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_gpu[python-lsp-server.test.Dockerfile-required_packages16] 
[gw5] [ 26%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_gpu[jupyterlab-git.test.Dockerfile-required_packages23] 
[gw1] [ 44%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_gpu[langchain.test.Dockerfile-required_packages26] 
[gw7] [ 53%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_gpu[jinja2.test.Dockerfile-required_packages36] 
[gw7] [ 55%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_gpu[fastapi.test.Dockerfile-required_packages38] 
[gw7] [ 61%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_gpu[uvicorn.test.Dockerfile-required_packages37] 
[gw6] [ 65%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_gpu[tf-keras.test.Dockerfile-required_packages43]
[gw7] [ 68%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_gpu[jupyter-scheduler.test.Dockerfile-required_packages40] 
[gw7] [ 70%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_gpu[ipywidgets.test.Dockerfile-required_packages41] 
[gw5] [ 72%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_gpu[supervisor.test.Dockerfile-required_packages42] 
[gw5] [ 74%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_gpu[jupyter-collaboration.test.Dockerfile-required_packages45]
[gw6] [ 80%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_gpu[jupyter-server-proxy.test.Dockerfile-required_packages46]
[gw5] [ 87%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_gpu[torchvision.test.Dockerfile-required_packages39] 
[gw0] [100%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_gpu[scikit-learn.test.Dockerfile-required_packages47] 
```
**Test results for v2.4.1-GPU**
```
[gw7] [  8%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_gpu[jupyterlab-lsp.test.Dockerfile-required_packages15] 
[gw1] [ 26%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_gpu[python-lsp-server.test.Dockerfile-required_packages16]
[gw6] [ 30%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_gpu[langchain.test.Dockerfile-required_packages26] 
[gw6] [ 32%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_gpu[jupyterlab-git.test.Dockerfile-required_packages23]
[gw7] [ 42%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_gpu[notebook.test.Dockerfile-required_packages18]
[gw5] [ 53%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_gpu[jinja2.test.Dockerfile-required_packages38]
[gw5] [ 55%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_gpu[fastapi.test.Dockerfile-required_packages40]
[gw5] [ 58%]PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_gpu[uvicorn.test.Dockerfile-required_packages39]
[gw3] [ 64%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_gpu[jupyter-scheduler.test.Dockerfile-required_packages42] 
[gw6] [ 65%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_gpu[ipywidgets.test.Dockerfile-required_packages43] 
[gw3] [ 67%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_gpu[tf-keras.test.Dockerfile-required_packages45] 
[gw6] [ 69%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_gpu[supervisor.test.Dockerfile-required_packages44] 
[gw6] [ 75%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_gpu[jupyter-collaboration.test.Dockerfile-required_packages47]
[gw3] [ 78%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_gpu[jupyter-collaboration.test.Dockerfile-required_packages47] 
[gw3] [ 80%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_gpu[scikit-learn.test.Dockerfile-required_packages49]
[gw7] [ 90%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_gpu[torchvision.test.Dockerfile-required_packages41]
```

**Test results for v3.0.0-GPU**
```
[gw7] [  8%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_gpu[jupyterlab-lsp.test.Dockerfile-required_packages15] 
[gw1] [ 27%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_gpu[python-lsp-server.test.Dockerfile-required_packages16]
[gw6] [ 31%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_gpu[langchain.test.Dockerfile-required_packages26]
[gw7] [ 39%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_gpu[notebook.test.Dockerfile-required_packages18] 
[gw6] [ 50%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_gpu[jinja2.test.Dockerfile-required_packages36] 
[gw6] [ 52%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_gpu[uvicorn.test.Dockerfile-required_packages37] 
[gw5] [ 58%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_gpu[ipywidgets.test.Dockerfile-required_packages41] 
[gw5] [ 66%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_gpu[supervisor.test.Dockerfile-required_packages42] 
[gw7] [ 72%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_gpu[fastapi.test.Dockerfile-required_packages38] 
[gw5] [ 75%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_gpu[tf-keras.test.Dockerfile-required_packages43] 
[gw5] [ 77%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_gpu[scikit-learn.test.Dockerfile-required_packages47] 
[gw3] [ 81%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_gpu[jupyter-collaboration.test.Dockerfile-required_packages45] 
[gw3] [ 83%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_gpu[jupyter-server-proxy.test.Dockerfile-required_packages46]
[gw6] [ 87%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_gpu[torchvision.test.Dockerfile-required_packages39] 
[gw6] [ 95%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_gpu[jupyter-scheduler.test.Dockerfile-required_packages40] 
[gw0] [ 97%] PASSED test/test_dockerfile_based_harness.py::test_dockerfiles_for_gpu[jupyterlab-git.test.Dockerfile-required_packages23] 
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/test_artifacts/v2/scripts/run_jupyterlab_git_tests.sh [hunk 1]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+
+pysdk_version=$(micromamba list | grep jupyter-git | tr -s ' ' | cut -d ' ' -f 3)
+
+git checkout tags/v$pysdk_version
```
Comment (aws-tianquaw) [inline test/test_artifacts/v2/scripts/run_jupyterlab_git_tests.sh:5]: nit: The variable name should not be "pysdk_version". Maybe "jupyter_git_version"

File: test/test_artifacts/v2/scripts/run_jupyterlab_lsp_tests.sh
Comment (aws-tianquaw) [inline test/test_artifacts/v2/scripts/run_jupyterlab_lsp_tests.sh:1]: For tests that are implemented in SMD repo, are they copied over from some other repository or you implemented them yourself?

**Review by aws-tianquaw — COMMENTED** [review]
Can you verify the new tests by running it against the latest patch version of each supported 2.x and 3.x image minor versions? Also attach the output of successful test runs to the PR.

In the PR description, also add a list of packages for which you added tests. This way we can better track when each test was added

File: test/test_artifacts/v2/fastapi.test.Dockerfile
[the remark this answers is not in the document]
Comment (bhavya2109sharma) [reply test/test_artifacts/v2/fastapi.test.Dockerfile:1]: Done

**Review by aws-tianquaw — DISMISSED** [review]

**Comment by claytonparnell** [comment]
I see example tests ran on v2 and v3, but only v2 test_artifacts are updated; do the v2 test_artifacts need to be copied over to v3?

**Review by aws-tianquaw — APPROVED** [review]

**Review by claytonparnell — APPROVED** [review]
